reps/2023-12-04-accelerated-dag.md

```diff
@@ -0,0 +1,637 @@
+from ray.dag import InputNode, MultiOutputNode
+
+a = Actor.remote()
+
+with InputNode() as inp:
+    dag = a.echo.bind(inp)
+
+# Produce a "compiled" DAG.
+dag = dag.experimental_compile()
+# Execute asynchronously. Return an ObjectRef, which can be `ray.get` and
+# `ray.wait` like normal ObjectRefs.
+ray.get(dag.execute(b"hello"))
+```
+
+## Tensor-parallel inference and training
+
+```python
+@ray.remote
+class ModelWorker:
+    def forward(self, inp):
+        # Internally, performs collective ops with other ModelWorkers.
+        return self.model.forward(inp)
+
+with InputNode() as inp:
+    dag = MultiOutputNode([worker.forward.bind(inp) for worker in workers])
+```
+
+## Pipeline-parallel inference
+
+Example applications:
+- Inter-node model parallelism
+- Pipelined inference between different models, e.g., target vs. draft models in speculative decoding
+
+```python
+@ray.remote
+class ModelWorker:
+    def forward(self, inp):
+        # Internally, performs collective ops with other ModelWorkers.
+        return self.model.forward(inp)
+
+
+with InputNode() as inp:
+    dag = inp
+    for worker in workers:
+        dag = worker.forward.bind(dag)
+```
+
+## Pipeline-parallel training
+
+Using [GPipe](https://arxiv.org/pdf/1811.06965.pdf)-style pipeline parallelism:
+
+```python
+T = 4  # Pipeline parallelism degree.
+workers = [ModelWorker.bind() for _ in range(T)]
+
+with InputNode() as inp:
+    microbatches_done = []
+    for microbatch in inp.microbatches:
+        for worker in workers:
+            microbatch = worker.forward.bind(microbatch)
+        for worker in reversed(workers):
+            microbatch = worker.backward.bind(microbatch)
+        # Empty objects signaling that the backwards pass is done.
+        microbatches_done.append(microbatch)
+    dag = MultiOutputNode([
+        worker.apply_gradient.bind(*microbatches_done)
+        for worker in workers])
+```
+
+## Transfering data with GPU channels
+
+Let's take an example where we want to overlap some execution on a producer/consumer with direct GPU-GPU transfer between them. One example of this is in [prefill disaggregation](https://arxiv.org/abs/2311.18677), where the prefill and decode forward passes for different layers may be overlapped with transfer of the KV cache for that layer.
+
+There are a few options for expressing this:
+
+1. Synchronously send data with the producer and consumer tasks. Simple, but GPUs will idle during the transfer.
+
+```python
+@ray.remote
+class Producer:
+    def produce_all_sync(self, inp):
+        gpu_tensors = []
+        for layer in model:
+            inp, gpu_tensor_shard = layer.forward(inp)
+            gpu_tensors.append(gpu_tensor_shard)
+
+        return inp, gpu_tensors
+
+@ray.remote
+class Consumer:
+    def consume_all_sync(self, gpu_tensors):
+        pass
+
+with InputNode() as inp:
+    with InputGPUChannel() as gpu_channel:
+        _, tensors = producer.produce_all_sync.bind(inp.input)
```

Comment:
This syntax is pretty strange, I'd expect something more like `tensors.set_device("gpu")` or `tensors.set_channel_impl(GPUChannel)`.

Reply:
Hmm, `tensors.set_channel_impl` seems good. Mainly I want there to be a way to pass a specific Channel or Channel type, so `tensors.set_device("gpu")` seems less ideal.